Query tips doc

writer-jill opened the pull request:
New topic containing tips for creating queries in Druid.

This PR has:
- [x] been self-reviewed.

ektravel commented on docs/querying/tips-good-queries.md:
```diff
@@ -0,0 +1,197 @@
+
+#### Use filters
+
+Use filters, for example the WHERE clause, and filter on time. Avoid using non-equal filters because they're very resource-intensive.
```
```suggestion
Use filters, such as the WHERE clause, and filter on time. Avoid using non-equal filters because they're very resource-intensive.
```

ektravel commented on docs/querying/tips-good-queries.md:
```diff
@@ -0,0 +1,197 @@
+The following example query demonstrates many of the tips outlined in this topic.
+The query:
+
+- selects specific dimensions and metrics.
```
```suggestion
- selects specific dimensions and metrics
```

ektravel commented on docs/querying/tips-good-queries.md:
```diff
@@ -0,0 +1,197 @@
+The query:
+
+- selects specific dimensions and metrics.
+- uses approximation.
```
```suggestion
- uses approximation
```

ektravel commented on docs/querying/tips-good-queries.md:
```diff
@@ -0,0 +1,197 @@
+
+- selects specific dimensions and metrics.
+- uses approximation.
+- selects from a single table.
```
```suggestion
- selects from a single table
```

ektravel commented on docs/querying/tips-good-queries.md:
```diff
@@ -0,0 +1,197 @@
+- selects specific dimensions and metrics.
+- uses approximation.
+- selects from a single table.
+- groups by low cardinality columns.
```
```suggestion
- groups by low cardinality columns
```

ektravel commented on docs/querying/tips-good-queries.md:
```diff
@@ -0,0 +1,197 @@
+- uses approximation.
+- selects from a single table.
+- groups by low cardinality columns.
+- filters on both dimensions and time.
```
```suggestion
- filters on both dimensions and time
```

ektravel commented on docs/querying/tips-good-queries.md:
```diff
@@ -0,0 +1,197 @@
+- selects from a single table.
+- groups by low cardinality columns.
+- filters on both dimensions and time.
+- orders by a dimension and a measure.
```
```suggestion
- orders by a dimension and a measure
```

ektravel commented on docs/querying/tips-good-queries.md:
```diff
@@ -0,0 +1,197 @@
+- groups by low cardinality columns.
+- filters on both dimensions and time.
+- orders by a dimension and a measure.
+- includes a limit.
```
```suggestion
- includes a limit
```

ektravel reviewed (approved):
The topic looks good. I removed punctuation from lines 178-184.

writer-jill commented:
> This is great! I'm glad to see this and all the new great resources it mentions too! A few possible, hopefully constructive, comments:
> 
> line 87 says "Avoid using non-equal filters because they're very resource-intensive." Then the example has a non-equal filter on time, which is normal (and an equal filter on product). Should it say something like "Avoid using non-equal filters when possible, because they're very resource-intensive"?
> 
> The query on lines 105/106 doesn't filter on time. Since we want to encourage that, should it add something like eg AND __time BETWEEN '2016-06-27 00:00' AND '2016-06-27 12:00' just to encourage filtering on time when possible?
> 
> For line 144, "Find out whether your dataset is already partitioned by user ID", I'm not sure what to do after I find that out, or how knowing helps or changes my options.
> 
> For line 146 and 148, "granularity" is a loaded word for druid (query granularity being something else then meant here). Maybe something like "Query over smaller intervals" and "Consider whether you can query a smaller time interval"?
> 
> line 194, is "__time" BETWEEN [] and [] on purpose? Should there be dates in there (like in my example above)?

Addressed all of these issues.

vtlim commented on docs/querying/tips-good-queries.md:
```diff
@@ -0,0 +1,198 @@
+
+Note that TopN queries are approximate in that each data process ranks its top K results and only returns those top K results to the Broker.
+
+You can follow the tutorial [Using TopN approximation in Druid queries](https://github.com/apache/druid/blob/master/examples/quickstart/jupyter-notebooks/notebooks/03-query/02-approxRanking.ipynb) to work through some examples with approximation turned on and off. The tutorial [Get to know query view](../tutorials/tutorial-sql-query-view.md) demonstrates running aggregate queries in the Druid con ...
```
```suggestion
You can follow the tutorial [Using TopN approximation in Druid queries](https://github.com/apache/druid/blob/master/examples/quickstart/jupyter-notebooks/notebooks/03-query/02-approxRanking.ipynb) to work through some examples with approximation turned on and off. The tutorial [Get to know Query view](../tutorials/tutorial-sql-query-view.md) demonstrates running aggregate queries in the Druid console.
```

vtlim commented on docs/querying/tips-good-queries.md:
```diff
@@ -0,0 +1,198 @@
+
+#### Use filters
+
+Use filters, for example the WHERE clause, and filter on time. Try to minimize the use of non-equal filters, because they're very resource-intensive.
```
"non-equal" -- maybe "inequality", if you're referring to lesser than or greater than

vtlim commented on docs/querying/tips-good-queries.md:
````diff
@@ -0,0 +1,198 @@
+```
+SELECT * from Wikipedia
+WHERE diffUrl LIKE 'https://en.wikipedia%'
+AND __time BETWEEN '2016-06-27 00:00' AND '2016-06-27 12:00'
````
The docs recommend against using SQL BETWEEN for time:
>Druid supports the standard SQL BETWEEN operator, but we recommend avoiding it for time filters. BETWEEN is inclusive of its upper bound, which makes it awkward to write time filters correctly. For example, the equivalent of TIME_IN_INTERVAL(__time, '2000-01-01/2000-02-01') is __time BETWEEN TIMESTAMP '2000-01-01 00:00:00' AND TIMESTAMP '2000-01-31 23:59:59.999'.

https://druid.apache.org/docs/latest/querying/sql-scalar#date-and-time-functions

Consider something like `TIME_IN_INTERVAL(__time, '2016-06-27/2016-06-27')`. However, when I tested this in a separate environment, I also needed `TIME_PARSE`.

vtlim replied:
Applies to example at bottom of doc as well

writer-jill replied:
Updated to an example that works for me - against Wikipedia sample data:
```
SELECT * from Wikipedia
WHERE diffUrl LIKE 'https://en.wikipedia%'
AND TIME_IN_INTERVAL(__time, '2016-06-27T01:00:00/2016-06-27T02:00:00')
```

writer-jill replied:
Also updated example at bottom of doc.

vtlim commented on docs/querying/tips-good-queries.md:
````diff
@@ -0,0 +1,198 @@
+GROUP BY id
+```
+
+To simply this query, you could split it into several queries and then aggregate the results. For example:
````
This example doesn't show the aggregation does it? Is that a manual step the user takes after running the following queries?

writer-jill replied:
Yes - updated wording to make this clearer.

vtlim commented on docs/querying/tips-good-queries.md:
```diff
@@ -0,0 +1,198 @@
+
+#### Minimize or remove subqueries
+
+Consider whether you can pre-compute a subquery task and store it as a join or make it a part of the datasource.
```
Further explanation, links to relevant docs, and/or examples would be helpful here

writer-jill replied:
Added a couple of links to more information and examples.

vtlim commented on docs/querying/tips-good-queries.md:
```diff
@@ -0,0 +1,198 @@
+
+Consider using Timeseries and TopN as alternatives to GroupBy. See [GroupBy queries: alternatives](./groupbyquery.md#alternatives) for more information.
+
+Avoid grouping on a high cardinality column, for example user ID. Investigate whether you can filter on user ID first. Find out whether your dataset is already partitioned by user ID. See [Partitioning](../ingestion/partitioning.md) for information on configuring paritioning.
```
This paragraph reads a bit odd. How do the points relate -- what is the takeaway if you CAN filter on user ID. If your dataset is partitioned, does that mean you can group on it?

Also typo in "partitioning" in the end of the sentence

writer-jill replied:
I don't think this part made much sense - removed the reference to partitioning here.

vtlim commented on docs/querying/tips-good-queries.md:
```diff
@@ -0,0 +1,198 @@
+
+#### Reduce the computation in your queries
+
+Examine your query to see if it uses a lot of transformation, functions, and expressions. Consider whether you could rewrite the query to reduce the level of computation.
```
```suggestion
Examine your query to see if it uses a lot of transformations, functions, and expressions. Consider whether you could rewrite the query to reduce the level of computation.
```
Consider an example here

vtlim commented on docs/querying/tips-good-queries.md:
````diff
@@ -0,0 +1,200 @@
+GROUP BY id
+```
+
+To simply this query, you could split it into several queries, for example:
````
```suggestion
To simplify this query, you could split it into several queries, for example:
```

vtlim reviewed (approved):
One nit. Otherwise looks good!
